src/manifest.c

```diff
@@ -382,166 +436,81 @@ int match_manifests(struct manifest *m1, struct manifest *m2)
+			 * triggers an attempt at creating a delta during pack
+			 * creation. For this case where the old file is deleted
+			 * or ghosted and is brought back we do not want this to
+			 * be attempted, as the old file will not be present
+			 * when trying to create the delta. */
+			if (!(file1->is_deleted && !file2->is_deleted) ||
+			    !(file1->is_ghosted && !file2->is_ghosted)) {
 				file1->peer = file2;
 				file2->peer = file1;
 			}
 
+			/* there was a match, advance both lists */
 			list1 = g_list_next(list1);
 			list2 = g_list_next(list2);
-			continue;
-		}
-		if (first) {
-			LOG(file1, "file added? ", "(file2 is %s)", file2->filename);
-			first = 0;
-		}
-		if (ret < 0) {
-			struct file *file3;
-			/*
-			 * if we get here, file1 got deleted... what we must do
-			 * is add a file entry for it in the target list.
-			 * However, since we're currently walking the list we
-			 * HAVE to prepend the entry.. and mark for sort at the
-			 * end.
-			 */
-			file3 = calloc(1, sizeof(struct file));
-			if (file3 == NULL) {
-				assert(0);
-			}
-
-			file3->filename = strdup(file1->filename);
-			hash_set_zeros(file3->hash);
-			file3->is_deleted = 1;
-			file3->is_config = file1->is_config;
-			file3->is_state = file1->is_state;
-			file3->is_boot = file1->is_boot;
-			file3->is_ghosted = file1->is_ghosted;
-			/* ghost deleted boot files */
-			if (!file3->is_ghosted) {
-				file3->is_ghosted = file1->is_boot && file1->is_deleted;
-			}
-
-			if (file3->is_ghosted || file1->is_deleted) {
-				/* if the new file is ghosted or the file was deleted, preserve
-				 * hash (all zeros if file1->is_deleted) and rename status */
-				file3->is_rename = file1->is_rename;
-				hash_assign(file1->hash, file3->hash);
-			}
-
-			/* for deleted files last_change remains the same.
-			 * otherwise last change is now */
-			file3->last_change = file1->is_deleted ? file1->last_change : m2->version;
-
-			file3->peer = file1;
-			file1->peer = file3;
-
-			list1 = g_list_next(list1);
-			m2->files = g_list_prepend(m2->files, file3);
-			m2->count++;
+		} else if (ret < 0) {
+			/* file1 was deleted, create entry for deleted file */
+			add_deleted_file(file1, m2);
 			if (!file1->is_deleted) {
 				account_deleted_file();
 				count++;
-				if (first) {
-					LOG(file1, "file got deleted", "");
-					first = 0;
-				}
 			}
+
 			must_sort = 1;
-			continue;
+			/* advance list1 for next file */
+			list1 = g_list_next(list1);
+		} else {
+			/* if we get here, ret is > 0, which means this is a new file added */
+			/* all we do is advance the pointer */
+			account_new_file();
+			count++;
+			/* advance list2 to check against same file in list1 */
+			list2 = g_list_next(list2);
 		}
-		/* if we get here, ret is > 0, which means this is a new file added */
-		/* all we do is advance the pointer */
-		account_new_file();
-		list2 = g_list_next(list2);
-		count++;
 	}
 
 	/* now deal with the tail ends */
 	while (list1) {
```

Comment:
Why? `while (something in list)` seems intuitive enough to me.

Reply:
I think there is a big difference between 'for each item in list' and 'while the list is not empty'. The latter is clearly a superset of the former.

So arguments for using for are
1. If you don't need the extra power that a while loop gives you, then you are conveying more accurately to your reader what you are wanting.
1. If we get to change these data structures to use arrays then we only need to fix the re-init in one place, rather than potentially many.
1. remember we have had bugs where people have had continue statements which lead to infinite loops because they have not moved on to the next statement.

As I said in my review, these changes are suggestions to make the code more readable, at least to me. They are just suggestions. At the end of the day it is your code.

Reply:
> If we get to change these data structures to use arrays then we only need to fix the re-init in one place, rather than potentially many.

this convinced me.